data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/s3/SqsWorker.java

```diff
@@ -336,6 +343,27 @@ private S3ObjectReference populateS3Reference(final String bucketName, final Str
                 .build();
     }
 
+    private ExpiryItem getExpiryItem(final Message message, final Instant expirationTime, final AcknowledgementSet acknowledgementSet) {
+        final long visibilityTimeoutSeconds = sqsOptions.getVisibilityTimeout().getSeconds();
+        return new ExpiryItem(message.messageId(), expirationTime.minusSeconds(visibilityTimeoutSeconds), visibilityTimeoutSeconds, expirationTime,
+                getExpiryCallback(message), acknowledgementSet);
+    }
+
+    private Consumer<ExpiryItem> getExpiryCallback(final Message message) {
+        return expiryItem -> {
+            final int newVisibilityTimeoutSeconds = (int) (expiryItem.getPollSeconds());
```

Comment:
We can also update this callback such that it notifies not only that the item is unexpired, but when it is completed or expired. For SQS, when this expires, the code can make a `ChangeMessageVisibility` request with the time set to `0`. This will make it immediately available for `ReceiveMessage`.

Reply:
That's a good idea